{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5c87ef3c-6dd0-4b09-ae7d-579e9da55163",
   "metadata": {},
   "source": [
    "### Table comparing barcoded Prom and Gridion\n",
    "Generates comparison tables for unblocked and sequenced metrics for the two tables.\n",
    "Goes from sequnecing summary files for these experiments, which are trimmed to only keep the columns we need, and are sorted based on the presence of the read_id in the `unblocked_read_ids.txt` from readfish.\n",
    "\n",
    "Note for me:\n",
    "fast5s for gridion run are here:\n",
    "`/mnt/wastewater/scratch_backup/ML_Runs_GridION/AP_002/BC-3-Panel-Human/20211025_1338_X1_FAQ39590_af835873/fast5*`\n",
    "\n",
    "fast5s for prom run are here:\n",
    "`/mnt/waterprom/ml_runs/PAS_001/NA12878_05_NB4_06_22Rv1_07`\n",
    "\n",
    "Sequencing summary files were created with guppy 6.4.6\n",
    "\n",
    "Separated with `rftools` which requires the unblocked read ids file, https://github.com/LooseLab/rftools\n",
    "\n",
    "```bash\n",
    "rftools split-ss -p <PREFIX> `20220317_1702_3A_PAK09329_4cae256c` sequencing_summary_PAK09329_e742ec3b.txt\n",
    "```\n",
    "\n",
    "where <PREFIX> is one of a combination of sequencing device and run number, i.e grid1, prom2 etc.\n",
    "\n",
    "Split sequencing summarys were trimmed to only contain relevant information using `qsvlite` - see https://github.com/jqnatividad/qsv?tab=readme-ov-file#option-3-install-with-rust\n",
    "Using the select command:\n",
    "```bash\n",
    "fd -e.txt ed -x qsvlite select -d \"\\t\" -o {.}_lite.tsv read_id,sample_id,sequence_length_template,barcode_arrangement {}\n",
    "```\n",
    "\n",
    "And then compressed with `xz` (in parallel with `fd`, just repalce {} with the filename to be compressed if not using `fd`\n",
    "\n",
    "```bash\n",
    "fd -tf lite -x xz -T 24 -e -z {}\n",
    "```\n",
    "\n",
    "### Alignment\n",
    "Alignment was performed using `minimap2 2.26-r1175`\n",
    "A script for alignment (`align_barcode_fastq`) is provided in the table 1 directory of the repository, the `map-ont` preset is used. This runs the alignment in Parallel for all files using 16 cores, so mayeb edit that.\n",
    "Both PASS and FAIL data is included.\n",
    "The data was called with HAC.\n",
    "\n",
    "### For coverage on and off target\n",
    "To generate files for on and off target coverage we used `mosdepth 0.3.6`.\n",
    "First the reads were mapped to Hg38.p14 with all alt and unplaced contigs removed, leaving only the main chromosome assemblies. \n",
    "Inverse complements regions (for off target based coverage) to the target bed files used were generated with `bedtools v2.31.1`\n",
    "This requires the bedfile to be sorted, this can be done with `bedtools sort`.\n",
    "\n",
    "```bash\n",
    "fd -e .sorted.bed -x sh -c 'bedtools complement -i -g /data/refs/hg38_simple.genome > {.}_inverse_complement.bed'\n",
    "```\n",
    "These are located in the target_beds.\n",
    "\n",
    "Mosdepth was used to calculate median coverage over the regions, and the median of this coverage was taken in this notebook. The targets were correct for each barcode. And example command:\n",
    "```bash\n",
    "mosdepth -m -t 32 -b CancerPanelTargets.sorted.bed -xn all_barcode_07_sequenced ../merged_reads/all_barcode07.sequenced.fastq.bam\n",
    "```\n",
    "\n",
    "An example script to run this in parallel for the Gridion run data is provided, `mosdepth_bam_files`. Again in parallel so user beware thread counts.\n",
    "\n",
    "##### Unclassified coverage\n",
    "For unclassified on/off target coverage, all three gene panel target bed files were merged, and overlaps were collapsed, but with strand awareness for the merging.\n",
    "Merging was performed with `bedtools v2.31.1` using the following command - \n",
    "```bash\n",
    "fd -ae sorted.bed -X cat {} | save -f all_targets.bed\n",
    "bedtools sort -i all_targets.bed | save -f all_targets.sorted.bed\n",
    "bedtools merge -s -i all_targets.sorted.bed | save -f all_targets.merged.sorted.bed\n",
    "bedtools complement -i all_targets.merged.sorted.bed -g /data/refs/hg38_simple.genome | save -f all_targets.sorted_inverse_complement.bed\n",
    "```\n",
    "\n",
    "The script for mosdepth also does the coverage for on/off target coverage in unclassified.\n",
    "\n",
    "## Warning\n",
    "This notebook is memory hungry (>20Gb RAM required). If this is an issue, contact us and we will work out a way to calculate these metrics without loading everything into a dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "id": "766bdc0d-a44a-4615-a5a4-70b2f1581609",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gc\n",
    "import re\n",
    "import sys\n",
    "from itertools import chain, repeat, zip_longest\n",
    "from pathlib import Path\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "# Define some constant values which are used in the table for information\n",
    "PAT = re.compile(r\"(sequenced|unblocked)\")\n",
    "gm, nb, rv = (\n",
    "    [\"GM12878\", \"TruSight 170 Tumour Panel\", 170],\n",
    "    [\"NB4\", \"TruSight RNA Fusion Panel\", 508],\n",
    "    [\"22Rv1\", \"COSMIC\", 717],\n",
    ")\n",
    "gridion_samples = {\n",
    "    \"barcode01\": gm,\n",
    "    \"barcode02\": nb,\n",
    "    \"barcode03\": rv,\n",
    "    \"unclassified_grid\": [\"*\", \"*\", \"*\"],\n",
    "}\n",
    "prom_samples = {\n",
    "    \"barcode05\": gm,\n",
    "    \"barcode06\": nb,\n",
    "    \"barcode07\": rv,\n",
    "    \"unclassified_prom\": [\"*\", \"*\", \"*\"],\n",
    "}\n",
    "# Combine which will be used in final table\n",
    "prom_samples.update(gridion_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8a0b9206-6e45-4c07-8254-67441425d76e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def n50(lengths: pd.Series) -> float:\n",
    "    \"\"\"\n",
    "    Returns the N50 of a series of read lengths\n",
    "    \"\"\"\n",
    "    lengths = lengths.sort_values(ascending=False).values\n",
    "    cumulative_sum = lengths.cumsum()\n",
    "    total_length = cumulative_sum[-1]\n",
    "    n50_length = lengths[(cumulative_sum >= total_length / 2).argmax()]\n",
    "    return n50_length"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36a090c0-b8a6-41e4-a28a-cde1b0d37a33",
   "metadata": {},
   "source": [
    "#### Read in all the xz compressed subset sequencing summary files for GridION experiment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6413e831-2c2b-41ed-baf0-fc1ed4fd8b4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "dfs = []\n",
    "for summary in Path(\".\").rglob(\"gridion*.tsv.xz\"):\n",
    "    read_type = PAT.findall(summary.name)[0]\n",
    "    df = pd.read_csv(summary, sep=\"\\t\")\n",
    "    df[\"read_type\"] = read_type\n",
    "    dfs.append(df)\n",
    "\n",
    "gridion_df = pd.concat(dfs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "id": "58b6c284-1704-4d26-b11c-ca576132de32",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1452772/3826483749.py:4: FutureWarning: The provided callable <function mean at 0x7f7f344cff40> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string \"mean\" instead.\n",
      "  ].agg([np.mean, np.sum, n50])\n",
      "/tmp/ipykernel_1452772/3826483749.py:4: FutureWarning: The provided callable <function sum at 0x7f7f344ceb00> is currently using SeriesGroupBy.sum. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string \"sum\" instead.\n",
      "  ].agg([np.mean, np.sum, n50])\n"
     ]
    }
   ],
   "source": [
    "# Calculate metrics for Gridion read lengths, yield and N50\n",
    "gridion_summary = gridion_df.groupby([\"read_type\", \"barcode_arrangement\"])[\n",
    "    \"sequence_length_template\"\n",
    "].agg([np.mean, np.sum, n50])\n",
    "# Only select relevant barcodes\n",
    "gridion_summary = gridion_summary.loc[\n",
    "    :, [\"barcode01\", \"barcode02\", \"barcode03\", \"unclassified\"], :\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99af8e68-de9b-462c-9657-8c904ab94e48",
   "metadata": {},
   "source": [
    "## Promethion\n",
    "#### Read in all the xz compressed subset sequencing summary files for PromethION experiment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e033d5ef-1c67-446c-b9a8-b711801ba91c",
   "metadata": {},
   "outputs": [],
   "source": [
    "dfs = []\n",
    "for summary in Path(\".\").rglob(\"sequencing_summaries/prom*.tsv.xz\"):\n",
    "    read_type = PAT.findall(summary.name)[0]\n",
    "    df = pd.read_csv(summary, sep=\"\\t\")\n",
    "    df[\"read_type\"] = read_type\n",
    "    dfs.append(df)\n",
    "\n",
    "prom_df = pd.concat(dfs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "03d3691d-39a6-44b8-8e7b-691e4de4c349",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1452772/4195199970.py:3: FutureWarning: The provided callable <function mean at 0x7f7f344cff40> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string \"mean\" instead.\n",
      "  ].agg([np.mean, np.sum, n50])\n",
      "/tmp/ipykernel_1452772/4195199970.py:3: FutureWarning: The provided callable <function sum at 0x7f7f344ceb00> is currently using SeriesGroupBy.sum. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string \"sum\" instead.\n",
      "  ].agg([np.mean, np.sum, n50])\n"
     ]
    }
   ],
   "source": [
    "prom_summary = prom_df.groupby([\"read_type\", \"barcode_arrangement\"])[\n",
    "    \"sequence_length_template\"\n",
    "].agg([np.mean, np.sum, n50])\n",
    "# Select correct barcodes\n",
    "prom_summary = prom_summary.loc[\n",
    "    :, [\"barcode05\", \"barcode06\", \"barcode07\", \"unclassified\"], :\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9a79e78a-3149-4103-a10c-84fa1bbd824d",
   "metadata": {},
   "outputs": [],
   "source": [
    "gridion_summary = gridion_summary.reset_index()\n",
    "gridion_summary[\"barcode_arrangement\"] = gridion_summary[\"barcode_arrangement\"].map(\n",
    "    lambda x: x if x != \"unclassified\" else \"unclassified_grid\"\n",
    ")\n",
    "gridion_summary[\"Device\"] = \"G.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "afb4d610-37f6-4fcd-a1de-b4e14f666ed4",
   "metadata": {},
   "outputs": [],
   "source": [
    "prom_summary = prom_summary.reset_index()\n",
    "prom_summary[\"barcode_arrangement\"] = prom_summary[\"barcode_arrangement\"].map(\n",
    "    lambda x: x if x != \"unclassified\" else \"unclassified_prom\"\n",
    ")\n",
    "prom_summary[\"Device\"] = \"P.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "79e95271-55d5-4fe1-acb1-d6823a2e8be6",
   "metadata": {},
   "outputs": [],
   "source": [
    "testing = pd.concat((gridion_summary, prom_summary))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "1d680980-0143-4ca9-a91a-e26712a31c11",
   "metadata": {},
   "outputs": [],
   "source": [
    "testing = testing.rename(\n",
    "    columns={\n",
    "        \"barcode_arrangement\": \"Barcode\",\n",
    "        \"mean\": \"Mean\",\n",
    "        \"sum\": \"Yield\",\n",
    "        \"n50\": \"N50\",\n",
    "    }\n",
    ")\n",
    "testing[\"read_type\"] = testing[\"read_type\"].map(\n",
    "    {\"sequenced\": \"Seq.\", \"unblocked\": \"Unb.\"}\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "id": "142d66b9-4c36-4a72-adb4-688dafb1478c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pivot from long form to wideform, keeping only relevant columns, more to be added later\n",
    "t = testing.pivot(\n",
    "    index=[\"Barcode\", \"Device\"],\n",
    "    columns=\"read_type\",\n",
    "    values=[\n",
    "        \"Yield\",\n",
    "        \"Mean\",\n",
    "        \"N50\",\n",
    "    ],\n",
    ")\n",
    "# Assign the panel information\n",
    "t[[\"Sample\", \"Panel\", \"Gene number\"]] = pd.DataFrame.from_records(\n",
    "    t.index.get_level_values(0).map(prom_samples), index=t.index\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6ddf391-a636-437d-83a3-ceb6ac1a77e9",
   "metadata": {},
   "source": [
    "#### Get the regions from the mosdepth files\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "id": "81ff5fb6-ae3e-4e45-b3dd-ccf8375c1037",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_coverages_of_regions() -> pd.DataFrame:\n",
    "    \"\"\"Read the mosdepth coverage region bed files filtering to the given barcodes.\n",
    "    Create a concatenated datafrom of all files, with barcode, coverage and on and off target and return it.\n",
    "    \"\"\"\n",
    "    prom = (\"prom\", \"05\", \"06\", \"07\", \"unclassified_prom\")\n",
    "    grid = (\"grid\", \"01\", \"02\", \"03\", \"unclassified_grid\")\n",
    "    pat = re.compile(r\"(barcode|unclassified[_promgrid]{5})_*(\\d{2})*.*\")\n",
    "    coverages = []\n",
    "    for barcode, device in chain(\n",
    "        zip_longest(prom[1:], [], fillvalue=prom[0]),\n",
    "        zip_longest(grid[1:], [], fillvalue=grid[0]),\n",
    "    ):\n",
    "        files = list(Path(\"mosdepth_region_summaries/\").glob(f\"*{barcode}*.bed.gz\"))\n",
    "        # print(files)\n",
    "        for region_file in files:\n",
    "            region_type = (\n",
    "                \"Off target\"\n",
    "                if \"off_target\" in str(region_file) or \"inverse\" in str(region_file)\n",
    "                else \"On target\"\n",
    "            )\n",
    "            print(f\"\\t{region_file}\\t{region_type}\")\n",
    "            barcode = \"\".join(pat.findall(str(region_file))[0])\n",
    "            # Unclassified and off target bed files do not provide names for records so we shift index for Coverage column down one\n",
    "            usecols = (\n",
    "                [3]\n",
    "                if region_type == \"Off target\" or \"unclassified\" in str(region_file)\n",
    "                else [4]\n",
    "            )\n",
    "            df = pd.read_csv(\n",
    "                region_file, sep=\"\\t\", header=None, usecols=usecols, names=[\"Coverage\"]\n",
    "            )\n",
    "            df[\"barcode\"] = barcode\n",
    "            df[\"region_type\"] = region_type\n",
    "            #             P. or G. to match the barcode/device assignement in main dataframe\n",
    "            df[\"Device\"] = f\"{device[0].upper()}.\"\n",
    "            coverages.append(df)\n",
    "    return pd.concat(coverages)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "id": "9b29a6b7-f339-4c28-b9f0-ad64bf38fd59",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tmosdepth_region_summaries/all_barcode_05.regions.bed.gz\tOn target\n",
      "\tmosdepth_region_summaries/all_barcode_05_off_target.regions.bed.gz\tOff target\n",
      "\tmosdepth_region_summaries/all_barcode_06.regions.bed.gz\tOn target\n",
      "\tmosdepth_region_summaries/all_barcode_06_off_target.regions.bed.gz\tOff target\n",
      "\tmosdepth_region_summaries/all_barcode_07.regions.bed.gz\tOn target\n",
      "\tmosdepth_region_summaries/all_barcode_07_off_target.regions.bed.gz\tOff target\n",
      "\tmosdepth_region_summaries/unclassified_prom_off_target.regions.bed.gz\tOff target\n",
      "\tmosdepth_region_summaries/unclassified_prom_on.regions.bed.gz\tOn target\n",
      "\tmosdepth_region_summaries/barcode01_tst_170_coords_extended.sorted.regions.bed.gz\tOn target\n",
      "\tmosdepth_region_summaries/barcode01_tst_170_coords_extended.sorted_inverse_complement.regions.bed.gz\tOff target\n",
      "\tmosdepth_region_summaries/barcode02_tst_fusion_coords_extended.sorted_inverse_complement.regions.bed.gz\tOff target\n",
      "\tmosdepth_region_summaries/barcode02_tst_fusion_coords_extended.sorted.regions.bed.gz\tOn target\n",
      "\tmosdepth_region_summaries/barcode03_CancerPanelTargets.sorted_inverse_complement.regions.bed.gz\tOff target\n",
      "\tmosdepth_region_summaries/barcode03_CancerPanelTargets.sorted.regions.bed.gz\tOn target\n",
      "\tmosdepth_region_summaries/unclassified_grid_all_targets.merged.sorted.regions.bed.gz\tOn target\n",
      "\tmosdepth_region_summaries/unclassified_grid_all_targets.sorted_inverse_complement.regions.bed.gz\tOff target\n"
     ]
    }
   ],
   "source": [
    "coverage_df = read_coverages_of_regions()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b8699ad-e2c1-4c97-8c18-945293985dad",
   "metadata": {},
   "source": [
    "#### Caluculate median coverage from the mosdepth files and get it into a format which can be ampped to amin dataframe via the index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "id": "2794dad3-4c70-4d36-9054-1fae0c3d00f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "a = (\n",
    "    coverage_df.groupby([\"barcode\", \"region_type\", \"Device\"])[\"Coverage\"]\n",
    "    .median()\n",
    "    .reset_index()\n",
    "    .pivot(index=[\"barcode\", \"Device\"], columns=\"region_type\", values=\"Coverage\")\n",
    ")\n",
    "a.index.names = [\"Barcode\", \"Device\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "id": "490e6a0d-6419-448f-b0a2-071e3ab8ecd4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1452772/2991135328.py:19: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '*' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.\n",
      "  t.loc[[\"unclassified_grid\", \"unclassified_prom\"], \"Fold Enrichment\"] = \"*\"\n"
     ]
    }
   ],
   "source": [
    "# Assign coverages\n",
    "t[[\"Off Target Median\", \"Target Median\"]] = a\n",
    "# Reorder the rows to this order\n",
    "t = t.reindex(\n",
    "    [\n",
    "        (\"barcode01\", \"G.\"),\n",
    "        (\"barcode02\", \"G.\"),\n",
    "        (\"barcode03\", \"G.\"),\n",
    "        (\"unclassified_grid\", \"G.\"),\n",
    "        (\"barcode05\", \"P.\"),\n",
    "        (\"barcode06\", \"P.\"),\n",
    "        (\"barcode07\", \"P.\"),\n",
    "        (\"unclassified_prom\", \"P.\"),\n",
    "    ]\n",
    ")\n",
    "# Assign absolute fold difference bewteen on and off target coverage\n",
    "t[\"Fold Enrichment\"] = t[\"Target Median\"] / t[\"Off Target Median\"]\n",
    "# Not used for the unclassified fields\n",
    "t.loc[[\"unclassified_grid\", \"unclassified_prom\"], \"Fold Enrichment\"] = \"*\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7151c95a-5fce-4547-af2e-87fc6b3f5885",
   "metadata": {},
   "source": [
    "#### Calculate the values for the total rows of the final table and assign them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "id": "251cef64-6f93-4d77-b95e-29eabf2ae8bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "total_df = (\n",
    "    t.groupby(level=1)\n",
    "    .agg(\n",
    "        {\n",
    "            (\"Mean\", \"Seq.\"): \"mean\",\n",
    "            (\"Mean\", \"Unb.\"): \"mean\",\n",
    "            (\"Yield\", \"Seq.\"): \"sum\",\n",
    "            (\"Yield\", \"Unb.\"): \"sum\",\n",
    "            (\"N50\", \"Seq.\"): \"mean\",\n",
    "            (\"N50\", \"Unb.\"): \"mean\",\n",
    "            (\"Off Target Median\", \"\"): \"median\",\n",
    "            (\"Target Median\", \"\"): \"median\",\n",
    "        }\n",
    "    )\n",
    "    .reset_index()\n",
    ")\n",
    "total_df[\"Barcode\"] = \"Total\"\n",
    "t = pd.concat((t, total_df.set_index([\"Barcode\", \"Device\"])))\n",
    "#  Reorder one more time\n",
    "t = t.reindex(\n",
    "    [\n",
    "        (\"barcode01\", \"G.\"),\n",
    "        (\"barcode02\", \"G.\"),\n",
    "        (\"barcode03\", \"G.\"),\n",
    "        (\"unclassified_grid\", \"G.\"),\n",
    "        (\"Total\", \"G.\"),\n",
    "        (\"barcode05\", \"P.\"),\n",
    "        (\"barcode06\", \"P.\"),\n",
    "        (\"barcode07\", \"P.\"),\n",
    "        (\"unclassified_prom\", \"P.\"),\n",
    "        (\"Total\", \"P.\"),\n",
    "    ]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "id": "5d8d657f-f92e-48a5-8c1c-18d1147f9fb8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th colspan=\"2\" halign=\"left\">Yield</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Mean</th>\n",
       "      <th colspan=\"2\" halign=\"left\">N50</th>\n",
       "      <th>Sample</th>\n",
       "      <th>Panel</th>\n",
       "      <th>Gene number</th>\n",
       "      <th>Off Target Median</th>\n",
       "      <th>Target Median</th>\n",
       "      <th>Fold Enrichment</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>read_type</th>\n",
       "      <th>Seq.</th>\n",
       "      <th>Unb.</th>\n",
       "      <th>Seq.</th>\n",
       "      <th>Unb.</th>\n",
       "      <th>Seq.</th>\n",
       "      <th>Unb.</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Barcode</th>\n",
       "      <th>Device</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>barcode01</th>\n",
       "      <th>G.</th>\n",
       "      <td>3.345244e+08</td>\n",
       "      <td>3.465450e+09</td>\n",
       "      <td>1926.417947</td>\n",
       "      <td>554.378839</td>\n",
       "      <td>8149.00</td>\n",
       "      <td>555.00</td>\n",
       "      <td>GM12878</td>\n",
       "      <td>TruSight 170 Tumour Panel</td>\n",
       "      <td>170</td>\n",
       "      <td>1.000</td>\n",
       "      <td>12.000</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcode02</th>\n",
       "      <th>G.</th>\n",
       "      <td>1.238984e+09</td>\n",
       "      <td>4.839475e+09</td>\n",
       "      <td>4203.349585</td>\n",
       "      <td>550.872480</td>\n",
       "      <td>7191.00</td>\n",
       "      <td>552.00</td>\n",
       "      <td>NB4</td>\n",
       "      <td>TruSight RNA Fusion Panel</td>\n",
       "      <td>508</td>\n",
       "      <td>1.000</td>\n",
       "      <td>15.000</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcode03</th>\n",
       "      <th>G.</th>\n",
       "      <td>1.248408e+09</td>\n",
       "      <td>3.835499e+09</td>\n",
       "      <td>5065.192309</td>\n",
       "      <td>555.972116</td>\n",
       "      <td>6858.00</td>\n",
       "      <td>556.00</td>\n",
       "      <td>22Rv1</td>\n",
       "      <td>COSMIC</td>\n",
       "      <td>717</td>\n",
       "      <td>1.000</td>\n",
       "      <td>12.000</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unclassified_grid</th>\n",
       "      <th>G.</th>\n",
       "      <td>1.709108e+08</td>\n",
       "      <td>3.658782e+09</td>\n",
       "      <td>798.171085</td>\n",
       "      <td>733.417091</td>\n",
       "      <td>923.00</td>\n",
       "      <td>792.00</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total</th>\n",
       "      <th>G.</th>\n",
       "      <td>2.992827e+09</td>\n",
       "      <td>1.579921e+10</td>\n",
       "      <td>2998.282731</td>\n",
       "      <td>598.660132</td>\n",
       "      <td>5780.25</td>\n",
       "      <td>613.75</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000</td>\n",
       "      <td>12.000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcode05</th>\n",
       "      <th>P.</th>\n",
       "      <td>1.278234e+09</td>\n",
       "      <td>1.476375e+10</td>\n",
       "      <td>930.660416</td>\n",
       "      <td>897.375697</td>\n",
       "      <td>7163.00</td>\n",
       "      <td>917.00</td>\n",
       "      <td>GM12878</td>\n",
       "      <td>TruSight 170 Tumour Panel</td>\n",
       "      <td>170</td>\n",
       "      <td>4.000</td>\n",
       "      <td>35.000</td>\n",
       "      <td>8.75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcode06</th>\n",
       "      <th>P.</th>\n",
       "      <td>4.357606e+09</td>\n",
       "      <td>2.339281e+10</td>\n",
       "      <td>1914.922736</td>\n",
       "      <td>903.235613</td>\n",
       "      <td>7349.00</td>\n",
       "      <td>919.00</td>\n",
       "      <td>NB4</td>\n",
       "      <td>TruSight RNA Fusion Panel</td>\n",
       "      <td>508</td>\n",
       "      <td>7.635</td>\n",
       "      <td>52.000</td>\n",
       "      <td>6.81074</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcode07</th>\n",
       "      <th>P.</th>\n",
       "      <td>3.010605e+09</td>\n",
       "      <td>1.363278e+10</td>\n",
       "      <td>2224.730111</td>\n",
       "      <td>909.816885</td>\n",
       "      <td>6999.00</td>\n",
       "      <td>923.00</td>\n",
       "      <td>22Rv1</td>\n",
       "      <td>COSMIC</td>\n",
       "      <td>717</td>\n",
       "      <td>4.000</td>\n",
       "      <td>27.250</td>\n",
       "      <td>6.8125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unclassified_prom</th>\n",
       "      <th>P.</th>\n",
       "      <td>7.036909e+08</td>\n",
       "      <td>1.549620e+10</td>\n",
       "      <td>507.391720</td>\n",
       "      <td>962.700204</td>\n",
       "      <td>543.00</td>\n",
       "      <td>989.00</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>4.000</td>\n",
       "      <td>5.000</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total</th>\n",
       "      <th>P.</th>\n",
       "      <td>9.350136e+09</td>\n",
       "      <td>6.728553e+10</td>\n",
       "      <td>1394.426246</td>\n",
       "      <td>918.282100</td>\n",
       "      <td>5513.50</td>\n",
       "      <td>937.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.000</td>\n",
       "      <td>31.125</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 Yield                       Mean              \\\n",
       "read_type                         Seq.          Unb.         Seq.        Unb.   \n",
       "Barcode           Device                                                        \n",
       "barcode01         G.      3.345244e+08  3.465450e+09  1926.417947  554.378839   \n",
       "barcode02         G.      1.238984e+09  4.839475e+09  4203.349585  550.872480   \n",
       "barcode03         G.      1.248408e+09  3.835499e+09  5065.192309  555.972116   \n",
       "unclassified_grid G.      1.709108e+08  3.658782e+09   798.171085  733.417091   \n",
       "Total             G.      2.992827e+09  1.579921e+10  2998.282731  598.660132   \n",
       "barcode05         P.      1.278234e+09  1.476375e+10   930.660416  897.375697   \n",
       "barcode06         P.      4.357606e+09  2.339281e+10  1914.922736  903.235613   \n",
       "barcode07         P.      3.010605e+09  1.363278e+10  2224.730111  909.816885   \n",
       "unclassified_prom P.      7.036909e+08  1.549620e+10   507.391720  962.700204   \n",
       "Total             P.      9.350136e+09  6.728553e+10  1394.426246  918.282100   \n",
       "\n",
       "                              N50           Sample                      Panel  \\\n",
       "read_type                    Seq.    Unb.                                       \n",
       "Barcode           Device                                                        \n",
       "barcode01         G.      8149.00  555.00  GM12878  TruSight 170 Tumour Panel   \n",
       "barcode02         G.      7191.00  552.00      NB4  TruSight RNA Fusion Panel   \n",
       "barcode03         G.      6858.00  556.00    22Rv1                     COSMIC   \n",
       "unclassified_grid G.       923.00  792.00        *                          *   \n",
       "Total             G.      5780.25  613.75      NaN                        NaN   \n",
       "barcode05         P.      7163.00  917.00  GM12878  TruSight 170 Tumour Panel   \n",
       "barcode06         P.      7349.00  919.00      NB4  TruSight RNA Fusion Panel   \n",
       "barcode07         P.      6999.00  923.00    22Rv1                     COSMIC   \n",
       "unclassified_prom P.       543.00  989.00        *                          *   \n",
       "Total             P.      5513.50  937.00      NaN                        NaN   \n",
       "\n",
       "                         Gene number Off Target Median Target Median  \\\n",
       "read_type                                                              \n",
       "Barcode           Device                                               \n",
       "barcode01         G.             170             1.000        12.000   \n",
       "barcode02         G.             508             1.000        15.000   \n",
       "barcode03         G.             717             1.000        12.000   \n",
       "unclassified_grid G.               *             0.000         1.000   \n",
       "Total             G.             NaN             1.000        12.000   \n",
       "barcode05         P.             170             4.000        35.000   \n",
       "barcode06         P.             508             7.635        52.000   \n",
       "barcode07         P.             717             4.000        27.250   \n",
       "unclassified_prom P.               *             4.000         5.000   \n",
       "Total             P.             NaN             4.000        31.125   \n",
       "\n",
       "                         Fold Enrichment  \n",
       "read_type                                 \n",
       "Barcode           Device                  \n",
       "barcode01         G.                12.0  \n",
       "barcode02         G.                15.0  \n",
       "barcode03         G.                12.0  \n",
       "unclassified_grid G.                   *  \n",
       "Total             G.                 NaN  \n",
       "barcode05         P.                8.75  \n",
       "barcode06         P.             6.81074  \n",
       "barcode07         P.              6.8125  \n",
       "unclassified_prom P.                   *  \n",
       "Total             P.                 NaN  "
      ]
     },
     "execution_count": 225,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t.head(12)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e168fbc2-6648-4693-9660-da1196ba79c9",
   "metadata": {},
   "source": [
    "#### Define some functions to format the cells for the latex table\n",
    "Otherwise cells are impossible to read"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "id": "11d8e7e4-c9fc-4bec-891e-f96e39d89d74",
   "metadata": {},
   "outputs": [],
   "source": [
    "def format_bases(num: int, factor: int = 1000, suffix: str = \"B\") -> str:\n",
    "    \"\"\"Return a human readable string of a large number using SI unit prefixes\n",
    "\n",
    "    :pararm num: A number to convert to decimal form\n",
    "    :param factor: The SI factor, use 1000 for SI units and 1024 for binary multiples\n",
    "    :param suffix: The suffix to place after the SI prefix, for example use B for SI units and iB for binary multiples\n",
    "    :return: The input number formatted to two decimal places with the SI unit and suffix\n",
    "\n",
    "    :Example:\n",
    "\n",
    "    >>> format_bases(1_000)\n",
    "    '1.00 kB'\n",
    "    >>> format_bases(1_000_000)\n",
    "    '1.00 MB'\n",
    "    >>> format_bases(1_630_000)\n",
    "    '1.63 MB'\n",
    "    >>> format_bases(1_000_000_000)\n",
    "    '1.00 GB'\n",
    "    \"\"\"\n",
    "    if isinstance(num, float) or isinstance(num, int):\n",
    "        for unit in [\"\", \"k\", \"M\", \"G\", \"T\", \"P\", \"E\", \"Z\"]:\n",
    "            if abs(num) < factor:\n",
    "                return f\"{num:3.2f}{unit}{suffix}\"\n",
    "            num /= factor\n",
    "        return f\"{num:3.2f}\"\n",
    "    else:\n",
    "        return num\n",
    "\n",
    "\n",
    "def format_fold(x: int) -> str:\n",
    "    \"\"\"\n",
    "    Format the fold change\n",
    "    \"\"\"\n",
    "    if not isinstance(x, float):\n",
    "        return x\n",
    "    # Check if the value is an integer\n",
    "    if isinstance(x, (int, float)) and x == int(x):\n",
    "        return f\"{int(x)}x\"\n",
    "    # Otherwise, format to one decimal place\n",
    "    elif isinstance(x, float):\n",
    "        return f\"{x:.1f}x\"\n",
    "    return x"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a64666ae-f158-4854-be9d-efa72e2afd07",
   "metadata": {},
   "source": [
    "#### Creat the latex table, and set up all the environments to make it look pretty.\n",
    "NB. I had to manually tweak some formatting the manuscript, such as making the MegaBase yields into 0.XXX (GB), and making the Total rows bold, and removing the End Reason cell and changing it to the word barcode"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "id": "8607e302-3f5a-4b17-a71e-ba891066e8a6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "t = t.fillna(\"\")\n",
    "t.columns.names = [None, \"End Reason\"]\n",
    "idx = pd.IndexSlice\n",
    "subset = idx[\n",
    "    :,\n",
    "    [(metric, end_reason) for end_reason in (\"Seq.\", \"Unb.\") for metric in [\"Yield\"]],\n",
    "]\n",
    "idx2 = pd.IndexSlice\n",
    "subset2 = idx2[:, (\"Fold Enrichment\", \"\")]\n",
    "idx3 = pd.IndexSlice\n",
    "subset3 = idx3[\n",
    "    :,\n",
    "    [\n",
    "        (\"Yield\", \"Seq.\"),\n",
    "        (\"Yield\", \"Unb.\"),\n",
    "        (\"N50\", \"Seq.\"),\n",
    "        (\"N50\", \"Unb.\"),\n",
    "        (\"Off Target Median\", \"\"),\n",
    "        (\"Target Median\", \"\"),\n",
    "        (\"Fold Enrichment\", \"\"),\n",
    "    ],\n",
    "]\n",
    "s = (\n",
    "    (\n",
    "        # Not proud of this loc, in here to exlude the mean columns, because I couldn't manage the slice to get the to hide\n",
    "        t.loc[:, ~t.columns.isin([(\"Mean\", \"Seq.\"), (\"Mean\", \"Unb.\")])]\n",
    "        .style.hide(axis=0, level=1, names=True)\n",
    "#          Format all numerical values to no decimal points\n",
    "        .format(precision=0, subset=subset3, thousands=\",\")\n",
    "        .format(format_bases, subset=subset)\n",
    "        # .format(\n",
    "        #     \"\\\\textbf{{{}}}\",\n",
    "        #     subset=idx[\n",
    "        #         [(\"Total\", \"G.\"), (\"Total\", \"P.\")],\n",
    "        #         :,\n",
    "        #     ],\n",
    "        #     precision=0,\n",
    "        #     thousands=\",\"\n",
    "        # )\n",
    "        .relabel_index(\n",
    "            t.loc[:, ~t.columns.isin([(\"Mean\", \"Seq.\"), (\"Mean\", \"Unb.\")])]\n",
    "            .columns.get_level_values(0)\n",
    "            .map(lambda x: \"Yield\\n(Gb)\" if str(x) == \"Yield\" else x),\n",
    "            axis=1,\n",
    "            level=0,\n",
    "        )\n",
    "        .format_index(escape=\"latex\", axis=0)\n",
    "#         Make columns label bold\n",
    "        .format_index(\"\\\\textbf{{{}}}\", escape=\"latex\", axis=1, level=[0, 1])\n",
    "        .to_latex(\n",
    "            position_float=\"centering\",\n",
    "            hrules=True,\n",
    "            column_format=\"\".join(repeat(\"X \", len(t.columns) - 2 + 1)),\n",
    "            environment=\"table\",\n",
    "        )\n",
    "    )\n",
    "#     Manual replaces of the latex string\n",
    "    .replace(r\"\\begin{tabular}\", r\"\\begin{tabularx}{\\textwidth}\")\n",
    "    .replace(\"\\centering\", \"\\centering\\n\\scriptsize\")\n",
    "    .replace(\"unclassified\\_grid\", \"unclassified\")\n",
    "    .replace(\"unclassified\\_prom\", \"unclassified\")\n",
    "    .replace(\"Total & 2.99\", \"\\midrule\\nTotal & 2.99\")\n",
    "    .replace(\"Total\", r\"\\textbf{Total}\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "id": "3e0a97ad-b7bb-497b-b3a4-1fa4842f7f25",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\\begin{table}\n",
      "\\centering\n",
      "\\scriptsize\n",
      "\\begin{tabularx}{\\textwidth}{X X X X X X X X X X X}\n",
      "\\toprule\n",
      " & \\multicolumn{2}{r}{\\textbf{Yield}} & \\multicolumn{2}{r}{\\textbf{N50}} & \\textbf{Sample} & \\textbf{Panel} & \\textbf{Gene number} & \\textbf{Off Target Median} & \\textbf{Target Median} & \\textbf{Fold Enrichment} \\\\\n",
      "End Reason & \\textbf{Seq.} & \\textbf{Unb.} & \\textbf{Seq.} & \\textbf{Unb.} & \\textbf{} & \\textbf{} & \\textbf{} & \\textbf{} & \\textbf{} & \\textbf{} \\\\\n",
      "\\midrule\n",
      "barcode01 & 334.52MB & 3.47GB & 8,149 & 555 & GM12878 & TruSight 170 Tumour Panel & 170 & 1 & 12 & 12 \\\\\n",
      "barcode02 & 1.24GB & 4.84GB & 7,191 & 552 & NB4 & TruSight RNA Fusion Panel & 508 & 1 & 15 & 15 \\\\\n",
      "barcode03 & 1.25GB & 3.84GB & 6,858 & 556 & 22Rv1 & COSMIC & 717 & 1 & 12 & 12 \\\\\n",
      "unclassified & 170.91MB & 3.66GB & 923 & 792 & * & * & * & 0 & 1 & * \\\\\n",
      "\\midrule\n",
      "\\textbf{Total} & 2.99GB & 15.80GB & 5,780 & 614 &  &  &  & 1 & 12 &  \\\\\n",
      "barcode05 & 1.28GB & 14.76GB & 7,163 & 917 & GM12878 & TruSight 170 Tumour Panel & 170 & 4 & 35 & 9 \\\\\n",
      "barcode06 & 4.36GB & 23.39GB & 7,349 & 919 & NB4 & TruSight RNA Fusion Panel & 508 & 8 & 52 & 7 \\\\\n",
      "barcode07 & 3.01GB & 13.63GB & 6,999 & 923 & 22Rv1 & COSMIC & 717 & 4 & 27 & 7 \\\\\n",
      "unclassified & 703.69MB & 15.50GB & 543 & 989 & * & * & * & 4 & 5 & * \\\\\n",
      "\\textbf{Total} & 9.35GB & 67.29GB & 5,514 & 937 &  &  &  & 4 & 31 &  \\\\\n",
      "\\bottomrule\n",
      "\\end{tabular}\n",
      "\\end{table}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f533b081-ae99-4365-91d8-7f601768d232",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
